diff --git a/arch/x86/boot/memory.c b/arch/x86/boot/memory.c
index d9c28c8..c0932a3 100644
--- a/arch/x86/boot/memory.c
+++ b/arch/x86/boot/memory.c
@@ -19,7 +19,7 @@
 
 static int detect_memory_e820(void)
 {
-	int count = 0;
+	unsigned int count = 0;
 	struct biosregs ireg, oreg;
 	struct boot_e820_entry *desc = boot_params.e820_table;
 	static struct boot_e820_entry buf; /* static so it is zeroed */
