diff --git a/arch/x86/events/intel/cqm.c b/arch/x86/events/intel/cqm.c
index 8c00dc0..52f5053 100644
--- a/arch/x86/events/intel/cqm.c
+++ b/arch/x86/events/intel/cqm.c
@@ -1469,7 +1469,7 @@ static struct attribute *intel_cmt_mbm_events_attr[] = {
 	NULL,
 };
 
-static struct attribute_group intel_cqm_events_group = {
+static attribute_group_no_const intel_cqm_events_group __read_only = {
 	.name = "events",
 	.attrs = NULL,
 };
@@ -1713,7 +1713,9 @@ static int __init intel_cqm_init(void)
 		goto out;
 	}
 
-	event_attr_intel_cqm_llc_scale.event_str = str;
+	pax_open_kernel();
+	const_cast(event_attr_intel_cqm_llc_scale.event_str) = str;
+	pax_close_kernel();
 
 	ret = intel_cqm_setup_rmid_cache();
 	if (ret)
@@ -1724,12 +1726,14 @@ static int __init intel_cqm_init(void)
 	if (ret && !cqm_enabled)
 		goto out;
 
+	pax_open_kernel();
 	if (cqm_enabled && mbm_enabled)
-		intel_cqm_events_group.attrs = intel_cmt_mbm_events_attr;
+		const_cast(intel_cqm_events_group.attrs) = intel_cmt_mbm_events_attr;
 	else if (!cqm_enabled && mbm_enabled)
-		intel_cqm_events_group.attrs = intel_mbm_events_attr;
+		const_cast(intel_cqm_events_group.attrs) = intel_mbm_events_attr;
 	else if (cqm_enabled && !mbm_enabled)
-		intel_cqm_events_group.attrs = intel_cqm_events_attr;
+		const_cast(intel_cqm_events_group.attrs) = intel_cqm_events_attr;
+	pax_close_kernel();
 
 	ret = perf_pmu_register(&intel_cqm_pmu, "intel_cqm", -1);
 	if (ret) {
