diff --git a/crypto/cryptd.c b/crypto/cryptd.c
index 0508c48..1eb1dd2 100644
--- a/crypto/cryptd.c
+++ b/crypto/cryptd.c
@@ -70,7 +70,7 @@ struct cryptd_blkcipher_ctx {
 
 struct cryptd_blkcipher_request_ctx {
 	crypto_completion_t complete;
-};
+} __no_const;
 
 struct cryptd_skcipher_ctx {
 	atomic_t refcnt;
@@ -79,7 +79,7 @@ struct cryptd_skcipher_ctx {
 
 struct cryptd_skcipher_request_ctx {
 	crypto_completion_t complete;
-};
+} __no_const;
 
 struct cryptd_hash_ctx {
 	atomic_t refcnt;
@@ -98,7 +98,7 @@ struct cryptd_aead_ctx {
 
 struct cryptd_aead_request_ctx {
 	crypto_completion_t complete;
-};
+} __no_const;
 
 static void cryptd_queue_worker(struct work_struct *work);
 
