diff --git a/Documentation/dontdiff b/Documentation/dontdiff
index 77b9222..ee2b836 100644
--- a/Documentation/dontdiff
+++ b/Documentation/dontdiff
@@ -109,6 +109,7 @@ config-*
 config_data.h*
 config.mak
 config.mak.autogen
+config.tmp
 conmakehash
 consolemap_deftbl.c*
 cpustr.h
@@ -120,9 +120,11 @@ devicetable-offsets.h
 dnotify_test
 dslm
 dtc
+dtc-lexer.lex.c
 elf2ecoff
 elfconfig.h*
 evergreen_reg_safe.h
+exception_policy.conf
 fixdep
 flask.h
 fore200e_mkfirm
@@ -150,14 +153,14 @@ int32.c
 int4.c
 int8.c
 kallsyms
-kconfig
+kern_constants.h
 keywords.c
 ksym.c*
 ksym.h*
 kxgettext
 *lex.c
 *lex.*.c
-linux
+lib1funcs.S
 logo_*.c
 logo_*_clut224.c
 logo_*_mono.c
@@ -217,9 +220,18 @@ series
 setup
 setup.bin
 setup.elf
+signing_key*
+aux.h
+disable.h
+e_fields.h
+e_fns.h
+e_fptrs.h
+e_vars.h
 sortextable
 sImage
+slabinfo
 sm_tbl*
+sortextable
 split-include
 syscalltab.h
 tables.c
