diff --git a/drivers/staging/ks7010/ks_wlan_net.c b/drivers/staging/ks7010/ks_wlan_net.c
index b2b4fa4..3120edc 100644
--- a/drivers/staging/ks7010/ks_wlan_net.c
+++ b/drivers/staging/ks7010/ks_wlan_net.c
@@ -174,9 +174,10 @@ int ks_wlan_setup_parameter(struct ks_wlan_private *priv,
  */
 
 static int ks_wlan_get_name(struct net_device *dev,
-			    struct iw_request_info *info, char *cwrq,
+			    struct iw_request_info *info, union iwreq_data *_cwrq,
 			    char *extra)
 {
+	char *cwrq = _cwrq->name;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -196,9 +197,10 @@ static int ks_wlan_get_name(struct net_device *dev,
 }
 
 static int ks_wlan_set_freq(struct net_device *dev,
-			    struct iw_request_info *info, struct iw_freq *fwrq,
+			    struct iw_request_info *info, union iwreq_data *_fwrq,
 			    char *extra)
 {
+	struct iw_freq *fwrq = &_fwrq->freq;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 	int channel;
 
@@ -240,9 +242,10 @@ static int ks_wlan_set_freq(struct net_device *dev,
 }
 
 static int ks_wlan_get_freq(struct net_device *dev,
-			    struct iw_request_info *info, struct iw_freq *fwrq,
+			    struct iw_request_info *info, union iwreq_data *_fwrq,
 			    char *extra)
 {
+	struct iw_freq *fwrq = &_fwrq->freq;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 	int f;
 
@@ -263,8 +266,9 @@ static int ks_wlan_get_freq(struct net_device *dev,
 
 static int ks_wlan_set_essid(struct net_device *dev,
 			     struct iw_request_info *info,
-			     struct iw_point *dwrq, char *extra)
+			     union iwreq_data *_dwrq, char *extra)
 {
+	struct iw_point *dwrq = &_dwrq->essid;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 	size_t len;
 
@@ -319,8 +323,9 @@ static int ks_wlan_set_essid(struct net_device *dev,
 
 static int ks_wlan_get_essid(struct net_device *dev,
 			     struct iw_request_info *info,
-			     struct iw_point *dwrq, char *extra)
+			     union iwreq_data *_dwrq, char *extra)
 {
+	struct iw_point *dwrq = &_dwrq->essid;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -348,8 +353,9 @@ static int ks_wlan_get_essid(struct net_device *dev,
 }
 
 static int ks_wlan_set_wap(struct net_device *dev, struct iw_request_info *info,
-			   struct sockaddr *ap_addr, char *extra)
+			   union iwreq_data *_ap_addr, char *extra)
 {
+	struct sockaddr *ap_addr = &_ap_addr->ap_addr;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	DPRINTK(2, "\n");
@@ -381,8 +387,9 @@ static int ks_wlan_set_wap(struct net_device *dev, struct iw_request_info *info,
 }
 
 static int ks_wlan_get_wap(struct net_device *dev, struct iw_request_info *info,
-			   struct sockaddr *awrq, char *extra)
+			   union iwreq_data *_awrq, char *extra)
 {
+	struct sockaddr *awrq = &_awrq->ap_addr;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -400,9 +407,10 @@ static int ks_wlan_get_wap(struct net_device *dev, struct iw_request_info *info,
 }
 
 static int ks_wlan_set_nick(struct net_device *dev,
-			    struct iw_request_info *info, struct iw_point *dwrq,
+			    struct iw_request_info *info, union iwreq_data *_dwrq,
 			    char *extra)
 {
+	struct iw_point *dwrq = &_dwrq->data;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -420,9 +428,10 @@ static int ks_wlan_set_nick(struct net_device *dev,
 }
 
 static int ks_wlan_get_nick(struct net_device *dev,
-			    struct iw_request_info *info, struct iw_point *dwrq,
+			    struct iw_request_info *info, union iwreq_data *_dwrq,
 			    char *extra)
 {
+	struct iw_point *dwrq = &_dwrq->data;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -437,9 +446,10 @@ static int ks_wlan_get_nick(struct net_device *dev,
 }
 
 static int ks_wlan_set_rate(struct net_device *dev,
-			    struct iw_request_info *info, struct iw_param *vwrq,
+			    struct iw_request_info *info, union iwreq_data *_vwrq,
 			    char *extra)
 {
+	struct iw_param *vwrq = &_vwrq->bitrate;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 	int i = 0;
 
@@ -669,9 +679,10 @@ static int ks_wlan_set_rate(struct net_device *dev,
 }
 
 static int ks_wlan_get_rate(struct net_device *dev,
-			    struct iw_request_info *info, struct iw_param *vwrq,
+			    struct iw_request_info *info, union iwreq_data *_vwrq,
 			    char *extra)
 {
+	struct iw_param *vwrq = &_vwrq->bitrate;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	DPRINTK(2, "in_interrupt = %ld update_phyinfo = %d\n",
@@ -694,8 +705,9 @@ static int ks_wlan_get_rate(struct net_device *dev,
 }
 
 static int ks_wlan_set_rts(struct net_device *dev, struct iw_request_info *info,
-			   struct iw_param *vwrq, char *extra)
+			   union iwreq_data *_vwrq, char *extra)
 {
+	struct iw_param *vwrq = &_vwrq->rts;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 	int rthr = vwrq->value;
 
@@ -715,8 +727,9 @@ static int ks_wlan_set_rts(struct net_device *dev, struct iw_request_info *info,
 }
 
 static int ks_wlan_get_rts(struct net_device *dev, struct iw_request_info *info,
-			   struct iw_param *vwrq, char *extra)
+			   union iwreq_data *_vwrq, char *extra)
 {
+	struct iw_param *vwrq = &_vwrq->rts;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -731,9 +744,10 @@ static int ks_wlan_get_rts(struct net_device *dev, struct iw_request_info *info,
 }
 
 static int ks_wlan_set_frag(struct net_device *dev,
-			    struct iw_request_info *info, struct iw_param *vwrq,
+			    struct iw_request_info *info, union iwreq_data *_vwrq,
 			    char *extra)
 {
+	struct iw_param *vwrq =&_vwrq->frag;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 	int fthr = vwrq->value;
 
@@ -754,9 +768,10 @@ static int ks_wlan_set_frag(struct net_device *dev,
 }
 
 static int ks_wlan_get_frag(struct net_device *dev,
-			    struct iw_request_info *info, struct iw_param *vwrq,
+			    struct iw_request_info *info, union iwreq_data *_vwrq,
 			    char *extra)
 {
+	struct iw_param *vwrq =&_vwrq->frag;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -771,9 +786,10 @@ static int ks_wlan_get_frag(struct net_device *dev,
 }
 
 static int ks_wlan_set_mode(struct net_device *dev,
-			    struct iw_request_info *info, __u32 *uwrq,
+			    struct iw_request_info *info, union iwreq_data *_uwrq,
 			    char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	DPRINTK(2, "mode=%d\n", *uwrq);
@@ -804,9 +820,10 @@ static int ks_wlan_set_mode(struct net_device *dev,
 }
 
 static int ks_wlan_get_mode(struct net_device *dev,
-			    struct iw_request_info *info, __u32 *uwrq,
+			    struct iw_request_info *info, union iwreq_data *_uwrq,
 			    char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -830,8 +847,9 @@ static int ks_wlan_get_mode(struct net_device *dev,
 
 static int ks_wlan_set_encode(struct net_device *dev,
 			      struct iw_request_info *info,
-			      struct iw_point *dwrq, char *extra)
+			      union iwreq_data *_dwrq, char *extra)
 {
+	struct iw_point *dwrq = &_dwrq->encoding;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	struct wep_key key;
@@ -945,8 +963,9 @@ static int ks_wlan_set_encode(struct net_device *dev,
 
 static int ks_wlan_get_encode(struct net_device *dev,
 			      struct iw_request_info *info,
-			      struct iw_point *dwrq, char *extra)
+			      union iwreq_data *_dwrq, char *extra)
 {
+	struct iw_point *dwrq = &_dwrq->encoding;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 	char zeros[16];
 	int index = (dwrq->flags & IW_ENCODE_INDEX) - 1;
@@ -998,15 +1017,16 @@ static int ks_wlan_get_encode(struct net_device *dev,
 #ifndef KSC_OPNOTSUPP
 static int ks_wlan_set_txpow(struct net_device *dev,
 			     struct iw_request_info *info,
-			     struct iw_param *vwrq, char *extra)
+			     union iwreq_data *vwrq, char *extra)
 {
 	return -EOPNOTSUPP;	/* Not Support */
 }
 
 static int ks_wlan_get_txpow(struct net_device *dev,
 			     struct iw_request_info *info,
-			     struct iw_param *vwrq, char *extra)
+			     union iwreq_data *_vwrq, char *extra)
 {
+	struct iw_param *vwrq = &_vwrq->txpower;
 	if (priv->sleep_mode == SLP_SLEEP)
 		return -EPERM;
 
@@ -1020,15 +1040,16 @@ static int ks_wlan_get_txpow(struct net_device *dev,
 
 static int ks_wlan_set_retry(struct net_device *dev,
 			     struct iw_request_info *info,
-			     struct iw_param *vwrq, char *extra)
+			     union iwreq_data *vwrq, char *extra)
 {
 	return -EOPNOTSUPP;	/* Not Support */
 }
 
 static int ks_wlan_get_retry(struct net_device *dev,
 			     struct iw_request_info *info,
-			     struct iw_param *vwrq, char *extra)
+			     union iwreq_data *_vwrq, char *extra)
 {
+	struct iw_param *vwrq = &_vwrq->txpower;
 	if (priv->sleep_mode == SLP_SLEEP)
 		return -EPERM;
 
@@ -1043,8 +1064,9 @@ static int ks_wlan_get_retry(struct net_device *dev,
 
 static int ks_wlan_get_range(struct net_device *dev,
 			     struct iw_request_info *info,
-			     struct iw_point *dwrq, char *extra)
+			     union iwreq_data *_dwrq, char *extra)
 {
+	struct iw_point *dwrq = &_dwrq->data;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 	struct iw_range *range = (struct iw_range *)extra;
 	int i, k;
@@ -1175,8 +1197,9 @@ static int ks_wlan_get_range(struct net_device *dev,
 
 static int ks_wlan_set_power(struct net_device *dev,
 			     struct iw_request_info *info,
-			     struct iw_param *vwrq, char *extra)
+			     union iwreq_data *_vwrq, char *extra)
 {
+	struct iw_param *vwrq =&_vwrq->power;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -1198,8 +1221,9 @@ static int ks_wlan_set_power(struct net_device *dev,
 
 static int ks_wlan_get_power(struct net_device *dev,
 			     struct iw_request_info *info,
-			     struct iw_param *vwrq, char *extra)
+			     union iwreq_data *_vwrq, char *extra)
 {
+	struct iw_param *vwrq =&_vwrq->power;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -1215,8 +1239,9 @@ static int ks_wlan_get_power(struct net_device *dev,
 
 static int ks_wlan_get_iwstats(struct net_device *dev,
 			       struct iw_request_info *info,
-			       struct iw_quality *vwrq, char *extra)
+			       union iwreq_data *_vwrq, char *extra)
 {
+	struct iw_quality *vwrq = &_vwrq->qual;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -1233,16 +1258,18 @@ static int ks_wlan_get_iwstats(struct net_device *dev,
 #ifndef KSC_OPNOTSUPP
 
 static int ks_wlan_set_sens(struct net_device *dev,
-			    struct iw_request_info *info, struct iw_param *vwrq,
+			    struct iw_request_info *info, union iwreq_data *vwrq,
 			    char *extra)
 {
 	return -EOPNOTSUPP;	/* Not Support */
 }
 
 static int ks_wlan_get_sens(struct net_device *dev,
-			    struct iw_request_info *info, struct iw_param *vwrq,
+			    struct iw_request_info *info, union iwreq_data *_vwrq,
 			    char *extra)
 {
+	struct iw_param *vwrq = &_vwrq->sens;
+
 	/* Not Support */
 	vwrq->value = 0;
 	vwrq->disabled = (vwrq->value == 0);
@@ -1254,8 +1281,9 @@ static int ks_wlan_get_sens(struct net_device *dev,
 /* Note : this is deprecated in favor of IWSCAN */
 static int ks_wlan_get_aplist(struct net_device *dev,
 			      struct iw_request_info *info,
-			      struct iw_point *dwrq, char *extra)
+			      union iwreq_data *_dwrq, char *extra)
 {
+	struct iw_point *dwrq = &_dwrq->data;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 	struct sockaddr *address = (struct sockaddr *)extra;
 	struct iw_quality qual[LOCAL_APLIST_MAX];
@@ -1478,9 +1506,10 @@ static inline char *ks_wlan_translate_scan(struct net_device *dev,
 }
 
 static int ks_wlan_get_scan(struct net_device *dev,
-			    struct iw_request_info *info, struct iw_point *dwrq,
+			    struct iw_request_info *info, union iwreq_data *_dwrq,
 			    char *extra)
 {
+	struct iw_point *dwrq = &_dwrq->data;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 	int i;
 	char *current_ev = extra;
@@ -1523,7 +1552,7 @@ static int ks_wlan_get_scan(struct net_device *dev,
 
 /* called after a bunch of SET operations */
 static int ks_wlan_config_commit(struct net_device *dev,
-				 struct iw_request_info *info, void *zwrq,
+				 struct iw_request_info *info, union iwreq_data *zwrq,
 				 char *extra)
 {
 	struct ks_wlan_private *priv = netdev_priv(dev);
@@ -1539,8 +1568,9 @@ static int ks_wlan_config_commit(struct net_device *dev,
 /* set association ie params */
 static int ks_wlan_set_genie(struct net_device *dev,
 			     struct iw_request_info *info,
-			     struct iw_point *dwrq, char *extra)
+			     union iwreq_data *_dwrq, char *extra)
 {
+	struct iw_point *dwrq =&_dwrq->data;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	DPRINTK(2, "\n");
@@ -1554,8 +1584,9 @@ static int ks_wlan_set_genie(struct net_device *dev,
 
 static int ks_wlan_set_auth_mode(struct net_device *dev,
 				 struct iw_request_info *info,
-				 struct iw_param *vwrq, char *extra)
+				 union iwreq_data *_vwrq, char *extra)
 {
+	struct iw_param *vwrq = &_vwrq->param;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 	int index = (vwrq->flags & IW_AUTH_INDEX);
 	int value = vwrq->value;
@@ -1688,8 +1719,9 @@ static int ks_wlan_set_auth_mode(struct net_device *dev,
 
 static int ks_wlan_get_auth_mode(struct net_device *dev,
 				 struct iw_request_info *info,
-				 struct iw_param *vwrq, char *extra)
+				 union iwreq_data *_vwrq, char *extra)
 {
+	struct iw_param *vwrq = &_vwrq->param;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 	int index = (vwrq->flags & IW_AUTH_INDEX);
 
@@ -1732,8 +1764,9 @@ static int ks_wlan_get_auth_mode(struct net_device *dev,
 /* set encoding token & mode (WPA)*/
 static int ks_wlan_set_encode_ext(struct net_device *dev,
 				  struct iw_request_info *info,
-				  struct iw_point *dwrq, char *extra)
+				  union iwreq_data *_dwrq, char *extra)
 {
+	struct iw_point *dwrq = &_dwrq->encoding;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 	struct iw_encode_ext *enc;
 	int index = dwrq->flags & IW_ENCODE_INDEX;
@@ -1828,8 +1861,9 @@ static int ks_wlan_set_encode_ext(struct net_device *dev,
 /* get encoding token & mode (WPA)*/
 static int ks_wlan_get_encode_ext(struct net_device *dev,
 				  struct iw_request_info *info,
-				  struct iw_point *dwrq, char *extra)
+				  union iwreq_data *_dwrq, char *extra)
 {
+	struct iw_point *dwrq = &_dwrq->encoding;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -1848,8 +1882,9 @@ static int ks_wlan_get_encode_ext(struct net_device *dev,
 
 static int ks_wlan_set_pmksa(struct net_device *dev,
 			     struct iw_request_info *info,
-			     struct iw_point *dwrq, char *extra)
+			     union iwreq_data *_dwrq, char *extra)
 {
+	struct iw_point *dwrq = &_dwrq->data;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 	struct iw_pmksa *pmksa;
 	int i;
@@ -1993,9 +2028,10 @@ static int ks_wlan_set_stop_request(struct net_device *dev,
 
 #include <linux/ieee80211.h>
 static int ks_wlan_set_mlme(struct net_device *dev,
-			    struct iw_request_info *info, struct iw_point *dwrq,
+			    struct iw_request_info *info, union iwreq_data *_dwrq,
 			    char *extra)
 {
+	struct iw_point *dwrq = &_dwrq->data;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 	struct iw_mlme *mlme = (struct iw_mlme *)extra;
 	__u32 mode;
@@ -2020,8 +2056,9 @@ static int ks_wlan_set_mlme(struct net_device *dev,
 
 static int ks_wlan_get_firmware_version(struct net_device *dev,
 					struct iw_request_info *info,
-					struct iw_point *dwrq, char *extra)
+					union iwreq_data *_dwrq, char *extra)
 {
+	struct iw_point *dwrq = &_dwrq->data;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	strcpy(extra, priv->firmware_version);
@@ -2030,9 +2067,10 @@ static int ks_wlan_get_firmware_version(struct net_device *dev,
 }
 
 static int ks_wlan_set_preamble(struct net_device *dev,
-				struct iw_request_info *info, __u32 *uwrq,
+				struct iw_request_info *info, union iwreq_data *_uwrq,
 				char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -2052,9 +2090,10 @@ static int ks_wlan_set_preamble(struct net_device *dev,
 }
 
 static int ks_wlan_get_preamble(struct net_device *dev,
-				struct iw_request_info *info, __u32 *uwrq,
+				struct iw_request_info *info, union iwreq_data *_uwrq,
 				char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -2066,9 +2105,10 @@ static int ks_wlan_get_preamble(struct net_device *dev,
 }
 
 static int ks_wlan_set_power_mgmt(struct net_device *dev,
-				  struct iw_request_info *info, __u32 *uwrq,
+				  struct iw_request_info *info, union iwreq_data *_uwrq,
 				  char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -2097,9 +2137,10 @@ static int ks_wlan_set_power_mgmt(struct net_device *dev,
 }
 
 static int ks_wlan_get_power_mgmt(struct net_device *dev,
-				  struct iw_request_info *info, __u32 *uwrq,
+				  struct iw_request_info *info, union iwreq_data *_uwrq,
 				  char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -2111,9 +2152,10 @@ static int ks_wlan_get_power_mgmt(struct net_device *dev,
 }
 
 static int ks_wlan_set_scan_type(struct net_device *dev,
-				 struct iw_request_info *info, __u32 *uwrq,
+				 struct iw_request_info *info, union iwreq_data *_uwrq,
 				 char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -2131,9 +2173,10 @@ static int ks_wlan_set_scan_type(struct net_device *dev,
 }
 
 static int ks_wlan_get_scan_type(struct net_device *dev,
-				 struct iw_request_info *info, __u32 *uwrq,
+				 struct iw_request_info *info, union iwreq_data *_uwrq,
 				 char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -2144,9 +2187,10 @@ static int ks_wlan_get_scan_type(struct net_device *dev,
 }
 
 static int ks_wlan_set_beacon_lost(struct net_device *dev,
-				   struct iw_request_info *info, __u32 *uwrq,
+				   struct iw_request_info *info, union iwreq_data *_uwrq,
 				   char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -2166,9 +2210,10 @@ static int ks_wlan_set_beacon_lost(struct net_device *dev,
 }
 
 static int ks_wlan_get_beacon_lost(struct net_device *dev,
-				   struct iw_request_info *info, __u32 *uwrq,
+				   struct iw_request_info *info, union iwreq_data *_uwrq,
 				   char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -2179,9 +2224,10 @@ static int ks_wlan_get_beacon_lost(struct net_device *dev,
 }
 
 static int ks_wlan_set_phy_type(struct net_device *dev,
-				struct iw_request_info *info, __u32 *uwrq,
+				struct iw_request_info *info, union iwreq_data *_uwrq,
 				char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -2202,9 +2248,10 @@ static int ks_wlan_set_phy_type(struct net_device *dev,
 }
 
 static int ks_wlan_get_phy_type(struct net_device *dev,
-				struct iw_request_info *info, __u32 *uwrq,
+				struct iw_request_info *info, union iwreq_data *_uwrq,
 				char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -2215,9 +2262,10 @@ static int ks_wlan_get_phy_type(struct net_device *dev,
 }
 
 static int ks_wlan_set_cts_mode(struct net_device *dev,
-				struct iw_request_info *info, __u32 *uwrq,
+				struct iw_request_info *info, union iwreq_data *_uwrq,
 				char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -2241,9 +2289,10 @@ static int ks_wlan_set_cts_mode(struct net_device *dev,
 }
 
 static int ks_wlan_get_cts_mode(struct net_device *dev,
-				struct iw_request_info *info, __u32 *uwrq,
+				struct iw_request_info *info, union iwreq_data *_uwrq,
 				char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -2255,8 +2304,9 @@ static int ks_wlan_get_cts_mode(struct net_device *dev,
 
 static int ks_wlan_set_sleep_mode(struct net_device *dev,
 				  struct iw_request_info *info,
-				  __u32 *uwrq, char *extra)
+				  union iwreq_data *_uwrq, char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	DPRINTK(2, "\n");
@@ -2282,8 +2332,9 @@ static int ks_wlan_set_sleep_mode(struct net_device *dev,
 
 static int ks_wlan_get_sleep_mode(struct net_device *dev,
 				  struct iw_request_info *info,
-				  __u32 *uwrq, char *extra)
+				  union iwreq_data *_uwrq, char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	DPRINTK(2, "GET_SLEEP_MODE %d\n", priv->sleep_mode);
@@ -2295,9 +2346,10 @@ static int ks_wlan_get_sleep_mode(struct net_device *dev,
 #ifdef WPS
 
 static int ks_wlan_set_wps_enable(struct net_device *dev,
-				  struct iw_request_info *info, __u32 *uwrq,
+				  struct iw_request_info *info, union iwreq_data *_uwrq,
 				  char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	DPRINTK(2, "\n");
@@ -2316,9 +2368,10 @@ static int ks_wlan_set_wps_enable(struct net_device *dev,
 }
 
 static int ks_wlan_get_wps_enable(struct net_device *dev,
-				  struct iw_request_info *info, __u32 *uwrq,
+				  struct iw_request_info *info, union iwreq_data *_uwrq,
 				  char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	DPRINTK(2, "\n");
@@ -2334,8 +2387,9 @@ static int ks_wlan_get_wps_enable(struct net_device *dev,
 
 static int ks_wlan_set_wps_probe_req(struct net_device *dev,
 				     struct iw_request_info *info,
-				     struct iw_point *dwrq, char *extra)
+				     union iwreq_data *_dwrq, char *extra)
 {
+	struct iw_point *dwrq = &_dwrq->data;
 	u8 *p = extra;
 	unsigned char len;
 	struct ks_wlan_private *priv = netdev_priv(dev);
@@ -2369,9 +2423,10 @@ static int ks_wlan_set_wps_probe_req(struct net_device *dev,
 #endif /* WPS */
 
 static int ks_wlan_set_tx_gain(struct net_device *dev,
-			       struct iw_request_info *info, __u32 *uwrq,
+			       struct iw_request_info *info, union iwreq_data *_uwrq,
 			       char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -2392,9 +2447,10 @@ static int ks_wlan_set_tx_gain(struct net_device *dev,
 }
 
 static int ks_wlan_get_tx_gain(struct net_device *dev,
-			       struct iw_request_info *info, __u32 *uwrq,
+			       struct iw_request_info *info, union iwreq_data *_uwrq,
 			       char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -2406,9 +2462,10 @@ static int ks_wlan_get_tx_gain(struct net_device *dev,
 }
 
 static int ks_wlan_set_rx_gain(struct net_device *dev,
-			       struct iw_request_info *info, __u32 *uwrq,
+			       struct iw_request_info *info, union iwreq_data *_uwrq,
 			       char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -2429,9 +2486,10 @@ static int ks_wlan_set_rx_gain(struct net_device *dev,
 }
 
 static int ks_wlan_get_rx_gain(struct net_device *dev,
-			       struct iw_request_info *info, __u32 *uwrq,
+			       struct iw_request_info *info, union iwreq_data *_uwrq,
 			       char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	if (priv->sleep_mode == SLP_SLEEP)
@@ -2443,9 +2501,10 @@ static int ks_wlan_get_rx_gain(struct net_device *dev,
 }
 
 static int ks_wlan_get_eeprom_cksum(struct net_device *dev,
-				    struct iw_request_info *info, __u32 *uwrq,
+				    struct iw_request_info *info, union iwreq_data *_uwrq,
 				    char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
 	*uwrq = priv->eeprom_checksum;
@@ -2567,8 +2626,9 @@ static void print_hif_event(struct net_device *dev, int event)
 
 /* get host command history */
 static int ks_wlan_hostt(struct net_device *dev, struct iw_request_info *info,
-			 __u32 *uwrq, char *extra)
+			 union iwreq_data *_uwrq, char *extra)
 {
+	__u32 *uwrq = &_uwrq->mode;
 	int i, event;
 	struct ks_wlan_private *priv = netdev_priv(dev);
 
@@ -2638,119 +2638,119 @@ static const struct iw_priv_args ks_wlan_private_args[] = {
 };
 
 static const iw_handler ks_wlan_handler[] = {
-	(iw_handler)ks_wlan_config_commit,	/* SIOCSIWCOMMIT */
-	(iw_handler)ks_wlan_get_name,	/* SIOCGIWNAME */
-	(iw_handler)NULL,	/* SIOCSIWNWID */
-	(iw_handler)NULL,	/* SIOCGIWNWID */
-	(iw_handler)ks_wlan_set_freq,	/* SIOCSIWFREQ */
-	(iw_handler)ks_wlan_get_freq,	/* SIOCGIWFREQ */
-	(iw_handler)ks_wlan_set_mode,	/* SIOCSIWMODE */
-	(iw_handler)ks_wlan_get_mode,	/* SIOCGIWMODE */
+	ks_wlan_config_commit,	/* SIOCSIWCOMMIT */
+	ks_wlan_get_name,	/* SIOCGIWNAME */
+	NULL,	/* SIOCSIWNWID */
+	NULL,	/* SIOCGIWNWID */
+	ks_wlan_set_freq,	/* SIOCSIWFREQ */
+	ks_wlan_get_freq,	/* SIOCGIWFREQ */
+	ks_wlan_set_mode,	/* SIOCSIWMODE */
+	ks_wlan_get_mode,	/* SIOCGIWMODE */
 #ifndef KSC_OPNOTSUPP
-	(iw_handler)ks_wlan_set_sens,	/* SIOCSIWSENS */
-	(iw_handler)ks_wlan_get_sens,	/* SIOCGIWSENS */
+	ks_wlan_set_sens,	/* SIOCSIWSENS */
+	ks_wlan_get_sens,	/* SIOCGIWSENS */
 #else /* KSC_OPNOTSUPP */
-	(iw_handler)NULL,	/* SIOCSIWSENS */
-	(iw_handler)NULL,	/* SIOCGIWSENS */
+	NULL,	/* SIOCSIWSENS */
+	NULL,	/* SIOCGIWSENS */
 #endif /* KSC_OPNOTSUPP */
-	(iw_handler)NULL,	/* SIOCSIWRANGE */
-	(iw_handler)ks_wlan_get_range,	/* SIOCGIWRANGE */
-	(iw_handler)NULL,	/* SIOCSIWPRIV */
-	(iw_handler)NULL,	/* SIOCGIWPRIV */
-	(iw_handler)NULL,	/* SIOCSIWSTATS */
-	(iw_handler)ks_wlan_get_iwstats,	/* SIOCGIWSTATS */
-	(iw_handler)NULL,	/* SIOCSIWSPY */
-	(iw_handler)NULL,	/* SIOCGIWSPY */
-	(iw_handler)NULL,	/* SIOCSIWTHRSPY */
-	(iw_handler)NULL,	/* SIOCGIWTHRSPY */
-	(iw_handler)ks_wlan_set_wap,	/* SIOCSIWAP */
-	(iw_handler)ks_wlan_get_wap,	/* SIOCGIWAP */
-//      (iw_handler)NULL,                      /* SIOCSIWMLME */
-	(iw_handler)ks_wlan_set_mlme,	/* SIOCSIWMLME */
-	(iw_handler)ks_wlan_get_aplist,	/* SIOCGIWAPLIST */
-	(iw_handler)ks_wlan_set_scan,	/* SIOCSIWSCAN */
-	(iw_handler)ks_wlan_get_scan,	/* SIOCGIWSCAN */
-	(iw_handler)ks_wlan_set_essid,	/* SIOCSIWESSID */
-	(iw_handler)ks_wlan_get_essid,	/* SIOCGIWESSID */
-	(iw_handler)ks_wlan_set_nick,	/* SIOCSIWNICKN */
-	(iw_handler)ks_wlan_get_nick,	/* SIOCGIWNICKN */
-	(iw_handler)NULL,	/* -- hole -- */
-	(iw_handler)NULL,	/* -- hole -- */
-	(iw_handler)ks_wlan_set_rate,	/* SIOCSIWRATE */
-	(iw_handler)ks_wlan_get_rate,	/* SIOCGIWRATE */
-	(iw_handler)ks_wlan_set_rts,	/* SIOCSIWRTS */
-	(iw_handler)ks_wlan_get_rts,	/* SIOCGIWRTS */
-	(iw_handler)ks_wlan_set_frag,	/* SIOCSIWFRAG */
-	(iw_handler)ks_wlan_get_frag,	/* SIOCGIWFRAG */
+	NULL,	/* SIOCSIWRANGE */
+	ks_wlan_get_range,	/* SIOCGIWRANGE */
+	NULL,	/* SIOCSIWPRIV */
+	NULL,	/* SIOCGIWPRIV */
+	NULL,	/* SIOCSIWSTATS */
+	ks_wlan_get_iwstats,	/* SIOCGIWSTATS */
+	NULL,	/* SIOCSIWSPY */
+	NULL,	/* SIOCGIWSPY */
+	NULL,	/* SIOCSIWTHRSPY */
+	NULL,	/* SIOCGIWTHRSPY */
+	ks_wlan_set_wap,	/* SIOCSIWAP */
+	ks_wlan_get_wap,	/* SIOCGIWAP */
+//      NULL,                      /* SIOCSIWMLME */
+	ks_wlan_set_mlme,	/* SIOCSIWMLME */
+	ks_wlan_get_aplist,	/* SIOCGIWAPLIST */
+	ks_wlan_set_scan,	/* SIOCSIWSCAN */
+	ks_wlan_get_scan,	/* SIOCGIWSCAN */
+	ks_wlan_set_essid,	/* SIOCSIWESSID */
+	ks_wlan_get_essid,	/* SIOCGIWESSID */
+	ks_wlan_set_nick,	/* SIOCSIWNICKN */
+	ks_wlan_get_nick,	/* SIOCGIWNICKN */
+	NULL,	/* -- hole -- */
+	NULL,	/* -- hole -- */
+	ks_wlan_set_rate,	/* SIOCSIWRATE */
+	ks_wlan_get_rate,	/* SIOCGIWRATE */
+	ks_wlan_set_rts,	/* SIOCSIWRTS */
+	ks_wlan_get_rts,	/* SIOCGIWRTS */
+	ks_wlan_set_frag,	/* SIOCSIWFRAG */
+	ks_wlan_get_frag,	/* SIOCGIWFRAG */
 #ifndef KSC_OPNOTSUPP
-	(iw_handler)ks_wlan_set_txpow,	/* SIOCSIWTXPOW */
-	(iw_handler)ks_wlan_get_txpow,	/* SIOCGIWTXPOW */
-	(iw_handler)ks_wlan_set_retry,	/* SIOCSIWRETRY */
-	(iw_handler)ks_wlan_get_retry,	/* SIOCGIWRETRY */
+	ks_wlan_set_txpow,	/* SIOCSIWTXPOW */
+	ks_wlan_get_txpow,	/* SIOCGIWTXPOW */
+	ks_wlan_set_retry,	/* SIOCSIWRETRY */
+	ks_wlan_get_retry,	/* SIOCGIWRETRY */
 #else /* KSC_OPNOTSUPP */
-	(iw_handler)NULL,	/* SIOCSIWTXPOW */
-	(iw_handler)NULL,	/* SIOCGIWTXPOW */
-	(iw_handler)NULL,	/* SIOCSIWRETRY */
-	(iw_handler)NULL,	/* SIOCGIWRETRY */
+	NULL,	/* SIOCSIWTXPOW */
+	NULL,	/* SIOCGIWTXPOW */
+	NULL,	/* SIOCSIWRETRY */
+	NULL,	/* SIOCGIWRETRY */
 #endif /* KSC_OPNOTSUPP */
-	(iw_handler)ks_wlan_set_encode,	/* SIOCSIWENCODE */
-	(iw_handler)ks_wlan_get_encode,	/* SIOCGIWENCODE */
-	(iw_handler)ks_wlan_set_power,	/* SIOCSIWPOWER */
-	(iw_handler)ks_wlan_get_power,	/* SIOCGIWPOWER */
-	(iw_handler)NULL,	/* -- hole -- */
-	(iw_handler)NULL,	/* -- hole -- */
-//      (iw_handler)NULL,                      /* SIOCSIWGENIE */
-	(iw_handler)ks_wlan_set_genie,	/* SIOCSIWGENIE */
-	(iw_handler)NULL,	/* SIOCGIWGENIE */
-	(iw_handler)ks_wlan_set_auth_mode,	/* SIOCSIWAUTH */
-	(iw_handler)ks_wlan_get_auth_mode,	/* SIOCGIWAUTH */
-	(iw_handler)ks_wlan_set_encode_ext,	/* SIOCSIWENCODEEXT */
-	(iw_handler)ks_wlan_get_encode_ext,	/* SIOCGIWENCODEEXT */
-	(iw_handler)ks_wlan_set_pmksa,	/* SIOCSIWPMKSA */
-	(iw_handler)NULL,	/* -- hole -- */
+	ks_wlan_set_encode,	/* SIOCSIWENCODE */
+	ks_wlan_get_encode,	/* SIOCGIWENCODE */
+	ks_wlan_set_power,	/* SIOCSIWPOWER */
+	ks_wlan_get_power,	/* SIOCGIWPOWER */
+	NULL,	/* -- hole -- */
+	NULL,	/* -- hole -- */
+//      NULL,                      /* SIOCSIWGENIE */
+	ks_wlan_set_genie,	/* SIOCSIWGENIE */
+	NULL,	/* SIOCGIWGENIE */
+	ks_wlan_set_auth_mode,	/* SIOCSIWAUTH */
+	ks_wlan_get_auth_mode,	/* SIOCGIWAUTH */
+	ks_wlan_set_encode_ext,	/* SIOCSIWENCODEEXT */
+	ks_wlan_get_encode_ext,	/* SIOCGIWENCODEEXT */
+	ks_wlan_set_pmksa,	/* SIOCSIWPMKSA */
+	NULL,	/* -- hole -- */
 };
 
 /* private_handler */
 static const iw_handler ks_wlan_private_handler[] = {
-	(iw_handler)NULL,	/*  0 */
-	(iw_handler)NULL,	/*  1, used to be: KS_WLAN_GET_DRIVER_VERSION */
-	(iw_handler)NULL,	/*  2 */
-	(iw_handler)ks_wlan_get_firmware_version,	/*  3 KS_WLAN_GET_FIRM_VERSION */
+	NULL,	/*  0 */
+	NULL,	/*  1, used to be: KS_WLAN_GET_DRIVER_VERSION */
+	NULL,	/*  2 */
+	ks_wlan_get_firmware_version,	/*  3 KS_WLAN_GET_FIRM_VERSION */
 #ifdef WPS
-	(iw_handler)ks_wlan_set_wps_enable,	/*  4 KS_WLAN_SET_WPS_ENABLE  */
-	(iw_handler)ks_wlan_get_wps_enable,	/*  5 KS_WLAN_GET_WPS_ENABLE  */
-	(iw_handler)ks_wlan_set_wps_probe_req,	/*  6 KS_WLAN_SET_WPS_PROBE_REQ */
+	ks_wlan_set_wps_enable,	/*  4 KS_WLAN_SET_WPS_ENABLE  */
+	ks_wlan_get_wps_enable,	/*  5 KS_WLAN_GET_WPS_ENABLE  */
+	ks_wlan_set_wps_probe_req,	/*  6 KS_WLAN_SET_WPS_PROBE_REQ */
 #else
-	(iw_handler)NULL,	/*  4 */
-	(iw_handler)NULL,	/*  5 */
-	(iw_handler)NULL,	/*  6 */
+	NULL,	/*  4 */
+	NULL,	/*  5 */
+	NULL,	/*  6 */
 #endif /* WPS */
 
-	(iw_handler)ks_wlan_get_eeprom_cksum,	/*  7 KS_WLAN_GET_CONNECT */
-	(iw_handler)ks_wlan_set_preamble,	/*  8 KS_WLAN_SET_PREAMBLE */
-	(iw_handler)ks_wlan_get_preamble,	/*  9 KS_WLAN_GET_PREAMBLE */
-	(iw_handler)ks_wlan_set_power_mgmt,	/* 10 KS_WLAN_SET_POWER_SAVE */
-	(iw_handler)ks_wlan_get_power_mgmt,	/* 11 KS_WLAN_GET_POWER_SAVE */
-	(iw_handler)ks_wlan_set_scan_type,	/* 12 KS_WLAN_SET_SCAN_TYPE */
-	(iw_handler)ks_wlan_get_scan_type,	/* 13 KS_WLAN_GET_SCAN_TYPE */
-	(iw_handler)ks_wlan_set_rx_gain,	/* 14 KS_WLAN_SET_RX_GAIN */
-	(iw_handler)ks_wlan_get_rx_gain,	/* 15 KS_WLAN_GET_RX_GAIN */
-	(iw_handler)ks_wlan_hostt,	/* 16 KS_WLAN_HOSTT */
-	(iw_handler)NULL,	/* 17 */
-	(iw_handler)ks_wlan_set_beacon_lost,	/* 18 KS_WLAN_SET_BECAN_LOST */
-	(iw_handler)ks_wlan_get_beacon_lost,	/* 19 KS_WLAN_GET_BECAN_LOST */
-	(iw_handler)ks_wlan_set_tx_gain,	/* 20 KS_WLAN_SET_TX_GAIN */
-	(iw_handler)ks_wlan_get_tx_gain,	/* 21 KS_WLAN_GET_TX_GAIN */
-	(iw_handler)ks_wlan_set_phy_type,	/* 22 KS_WLAN_SET_PHY_TYPE */
-	(iw_handler)ks_wlan_get_phy_type,	/* 23 KS_WLAN_GET_PHY_TYPE */
-	(iw_handler)ks_wlan_set_cts_mode,	/* 24 KS_WLAN_SET_CTS_MODE */
-	(iw_handler)ks_wlan_get_cts_mode,	/* 25 KS_WLAN_GET_CTS_MODE */
-	(iw_handler)NULL,	/* 26 */
-	(iw_handler)NULL,	/* 27 */
-	(iw_handler)ks_wlan_set_sleep_mode,	/* 28 KS_WLAN_SET_SLEEP_MODE */
-	(iw_handler)ks_wlan_get_sleep_mode,	/* 29 KS_WLAN_GET_SLEEP_MODE */
-	(iw_handler)NULL,	/* 30 */
-	(iw_handler)NULL,	/* 31 */
+	ks_wlan_get_eeprom_cksum,	/*  7 KS_WLAN_GET_CONNECT */
+	ks_wlan_set_preamble,	/*  8 KS_WLAN_SET_PREAMBLE */
+	ks_wlan_get_preamble,	/*  9 KS_WLAN_GET_PREAMBLE */
+	ks_wlan_set_power_mgmt,	/* 10 KS_WLAN_SET_POWER_SAVE */
+	ks_wlan_get_power_mgmt,	/* 11 KS_WLAN_GET_POWER_SAVE */
+	ks_wlan_set_scan_type,	/* 12 KS_WLAN_SET_SCAN_TYPE */
+	ks_wlan_get_scan_type,	/* 13 KS_WLAN_GET_SCAN_TYPE */
+	ks_wlan_set_rx_gain,	/* 14 KS_WLAN_SET_RX_GAIN */
+	ks_wlan_get_rx_gain,	/* 15 KS_WLAN_GET_RX_GAIN */
+	ks_wlan_hostt,	/* 16 KS_WLAN_HOSTT */
+	NULL,	/* 17 */
+	ks_wlan_set_beacon_lost,	/* 18 KS_WLAN_SET_BECAN_LOST */
+	ks_wlan_get_beacon_lost,	/* 19 KS_WLAN_GET_BECAN_LOST */
+	ks_wlan_set_tx_gain,	/* 20 KS_WLAN_SET_TX_GAIN */
+	ks_wlan_get_tx_gain,	/* 21 KS_WLAN_GET_TX_GAIN */
+	ks_wlan_set_phy_type,	/* 22 KS_WLAN_SET_PHY_TYPE */
+	ks_wlan_get_phy_type,	/* 23 KS_WLAN_GET_PHY_TYPE */
+	ks_wlan_set_cts_mode,	/* 24 KS_WLAN_SET_CTS_MODE */
+	ks_wlan_get_cts_mode,	/* 25 KS_WLAN_GET_CTS_MODE */
+	NULL,	/* 26 */
+	NULL,	/* 27 */
+	ks_wlan_set_sleep_mode,	/* 28 KS_WLAN_SET_SLEEP_MODE */
+	ks_wlan_get_sleep_mode,	/* 29 KS_WLAN_GET_SLEEP_MODE */
+	NULL,	/* 30 */
+	NULL,	/* 31 */
 };
 
 static const struct iw_handler_def ks_wlan_handler_def = {
@@ -2758,9 +2758,9 @@ static const struct iw_handler_def ks_wlan_handler_def = {
 	.num_private = sizeof(ks_wlan_private_handler) / sizeof(iw_handler),
 	.num_private_args =
 	    sizeof(ks_wlan_private_args) / sizeof(struct iw_priv_args),
-	.standard = (iw_handler *)ks_wlan_handler,
-	.private = (iw_handler *)ks_wlan_private_handler,
-	.private_args = (struct iw_priv_args *)ks_wlan_private_args,
+	.standard = ks_wlan_handler,
+	.private = ks_wlan_private_handler,
+	.private_args = ks_wlan_private_args,
 	.get_wireless_stats = ks_get_wireless_stats,
 };
 
@@ -2825,7 +2825,7 @@ void ks_wlan_tx_timeout(struct net_device *dev)
 }
 
 static
-int ks_wlan_start_xmit(struct sk_buff *skb, struct net_device *dev)
+netdev_tx_t ks_wlan_start_xmit(struct sk_buff *skb, struct net_device *dev)
 {
 	struct ks_wlan_private *priv = netdev_priv(dev);
 	int ret;
