diff --git a/drivers/staging/lustre/lustre/lov/lov_request.c b/drivers/staging/lustre/lustre/lov/lov_request.c
index 09dcaf4..8d02a46 100644
--- a/drivers/staging/lustre/lustre/lov/lov_request.c
+++ b/drivers/staging/lustre/lustre/lov/lov_request.c
@@ -41,8 +41,8 @@
 static void lov_init_set(struct lov_request_set *set)
 {
 	set->set_count = 0;
-	atomic_set(&set->set_completes, 0);
-	atomic_set(&set->set_success, 0);
+	atomic_set_unchecked(&set->set_completes, 0);
+	atomic_set_unchecked(&set->set_success, 0);
 	INIT_LIST_HEAD(&set->set_list);
 }
 
@@ -66,9 +66,9 @@ static void lov_finish_set(struct lov_request_set *set)
 static void lov_update_set(struct lov_request_set *set,
 			   struct lov_request *req, int rc)
 {
-	atomic_inc(&set->set_completes);
+	atomic_inc_unchecked(&set->set_completes);
 	if (rc == 0)
-		atomic_inc(&set->set_success);
+		atomic_inc_unchecked(&set->set_success);
 }
 
 static void lov_set_add_req(struct lov_request *req,
@@ -178,9 +178,9 @@ int lov_fini_statfs_set(struct lov_request_set *set)
 	if (!set)
 		return 0;
 
-	if (atomic_read(&set->set_completes)) {
+	if (atomic_read_unchecked(&set->set_completes)) {
 		rc = lov_fini_statfs(set->set_obd, set->set_oi->oi_osfs,
-				     atomic_read(&set->set_success));
+				     atomic_read_unchecked(&set->set_success));
 	}
 
 	lov_finish_set(set);
@@ -576,7 +576,7 @@ static int cb_statfs_update(void *cookie, int rc)
 	lov = &lovobd->u.lov;
 	osfs = set->set_oi->oi_osfs;
 	lov_sfs = oinfo->oi_osfs;
-	success = atomic_read(&set->set_success);
+	success = atomic_read_unchecked(&set->set_success);
 	/* XXX: the same is done in lov_update_common_set, however
 	 * lovset->set_exp is not initialized.
 	 */
