diff --git a/drivers/net/can/vcan.c b/drivers/net/can/vcan.c
index a8cb332..e4c774d 100644
--- a/drivers/net/can/vcan.c
+++ b/drivers/net/can/vcan.c
@@ -166,7 +166,7 @@ static void vcan_setup(struct net_device *dev)
 	dev->needs_free_netdev	= true;
 }
 
-static struct rtnl_link_ops vcan_link_ops __read_mostly = {
+static struct rtnl_link_ops vcan_link_ops = {
 	.kind	= DRV_NAME,
 	.setup	= vcan_setup,
 };
