diff --git a/drivers/scsi/qla2xxx/qla_target.h b/drivers/scsi/qla2xxx/qla_target.h
index d644202..660cf9f 100644
--- a/drivers/scsi/qla2xxx/qla_target.h
+++ b/drivers/scsi/qla2xxx/qla_target.h
@@ -821,7 +821,7 @@ struct qla_tgt {
 	int notify_ack_expected;
 	int abts_resp_expected;
 	int modify_lun_expected;
-	atomic_t tgt_global_resets_count;
+	atomic_unchecked_t tgt_global_resets_count;
 	struct list_head tgt_list_entry;
 };
 
