diff --git a/drivers/staging/lustre/lustre/include/llog_swab.h b/drivers/staging/lustre/lustre/include/llog_swab.h
index fd7ffb1..d8e3b20 100644
--- a/drivers/staging/lustre/lustre/include/llog_swab.h
+++ b/drivers/staging/lustre/lustre/include/llog_swab.h
@@ -52,15 +52,15 @@
 
 struct lustre_cfg;
 
-void lustre_swab_lu_fid(struct lu_fid *fid);
-void lustre_swab_ost_id(struct ost_id *oid);
-void lustre_swab_llogd_body(struct llogd_body *d);
-void lustre_swab_llog_hdr(struct llog_log_hdr *h);
-void lustre_swab_llogd_conn_body(struct llogd_conn_body *d);
-void lustre_swab_llog_rec(struct llog_rec_hdr *rec);
-void lustre_swab_lu_seq_range(struct lu_seq_range *range);
-void lustre_swab_lustre_cfg(struct lustre_cfg *lcfg);
-void lustre_swab_cfg_marker(struct cfg_marker *marker,
+void lustre_swab_lu_fid(void *fid);
+void lustre_swab_ost_id(void *oid);
+void lustre_swab_llogd_body(void *d);
+void lustre_swab_llog_hdr(void *h);
+void lustre_swab_llogd_conn_body(void *d);
+void lustre_swab_llog_rec(void *rec);
+void lustre_swab_lu_seq_range(void *range);
+void lustre_swab_lustre_cfg(void *lcfg);
+void lustre_swab_cfg_marker(void *marker,
 			    int swab, int size);
 
 #endif
