diff --git a/drivers/net/ethernet/amd/xgbe/xgbe.h b/drivers/net/ethernet/amd/xgbe/xgbe.h
index 5dd17dc..962ff6a 100644
--- a/drivers/net/ethernet/amd/xgbe/xgbe.h
+++ b/drivers/net/ethernet/amd/xgbe/xgbe.h
@@ -994,10 +994,10 @@ struct xgbe_prv_data {
 
 	char ecc_name[IFNAMSIZ + 32];
 
-	struct xgbe_hw_if hw_if;
-	struct xgbe_phy_if phy_if;
-	struct xgbe_desc_if desc_if;
-	struct xgbe_i2c_if i2c_if;
+	const struct xgbe_hw_if *hw_if;
+	const struct xgbe_phy_if *phy_if;
+	const struct xgbe_desc_if *desc_if;
+	const struct xgbe_i2c_if *i2c_if;
 
 	/* AXI DMA settings */
 	unsigned int coherent;
@@ -1173,6 +1173,11 @@ struct xgbe_prv_data {
 #endif
 };
 
+extern const struct xgbe_hw_if default_xgbe_hw_if;
+extern const struct xgbe_phy_if default_xgbe_phy_if;
+extern const struct xgbe_desc_if default_xgbe_desc_if;
+extern const struct xgbe_i2c_if default_xgbe_i2c_if;
+
 /* Function prototypes*/
 struct xgbe_prv_data *xgbe_alloc_pdata(struct device *);
 void xgbe_free_pdata(struct xgbe_prv_data *);
