diff --git a/fs/ceph/super.c b/fs/ceph/super.c
index b382e59..e12e43f 100644
--- a/fs/ceph/super.c
+++ b/fs/ceph/super.c
@@ -935,14 +935,14 @@ static int ceph_compare_super(struct super_block *sb, void *data)
 /*
  * construct our own bdi so we can control readahead, etc.
  */
-static atomic_long_t bdi_seq = ATOMIC_LONG_INIT(0);
+static atomic_long_unchecked_t bdi_seq = ATOMIC_LONG_INIT(0);
 
 static int ceph_setup_bdi(struct super_block *sb, struct ceph_fs_client *fsc)
 {
 	int err;
 
 	err = super_setup_bdi_name(sb, "ceph-%ld",
-				   atomic_long_inc_return(&bdi_seq));
+				   atomic_long_inc_return_unchecked(&bdi_seq));
 	if (err)
 		return err;
 
