diff --git a/fs/read_write.c b/fs/read_write.c
index 190e0d36..2ac0f9c 100644
--- a/fs/read_write.c
+++ b/fs/read_write.c
@@ -420,10 +420,13 @@ ssize_t kernel_read(struct file *file, void *buf, size_t count, loff_t *pos)
 	mm_segment_t old_fs;
 	ssize_t result;
 
+	if (count > INT_MAX)
+		return -EINVAL;
+
 	old_fs = get_fs();
 	set_fs(get_ds());
 	/* The cast to a user pointer is valid due to the set_fs() */
-	result = vfs_read(file, (void __user *)buf, count, pos);
+	result = vfs_read(file, (void __force_user *)buf, count, pos);
 	set_fs(old_fs);
 	return result;
 }
@@ -495,7 +495,7 @@ ssize_t __kernel_write(struct file *file, const void *buf, size_t count, loff_t
 
 	old_fs = get_fs();
 	set_fs(get_ds());
-	p = (__force const char __user *)buf;
+	p = (const char __force_user *)buf;
 	if (count > MAX_RW_COUNT)
 		count =  MAX_RW_COUNT;
 	ret = __vfs_write(file, p, count, pos);
@@ -518,7 +521,7 @@ ssize_t kernel_write(struct file *file, const void *buf, size_t count,
 	old_fs = get_fs();
 	set_fs(get_ds());
 	/* The cast to a user pointer is valid due to the set_fs() */
-	res = vfs_write(file, (__force const char __user *)buf, count, pos);
+	res = vfs_write(file, (const char __force_user *)buf, count, pos);
 	set_fs(old_fs);
 
 	return res;
