diff --git a/fs/filesystems.c b/fs/filesystems.c
index c5618db..50c38f4 100644
--- a/fs/filesystems.c
+++ b/fs/filesystems.c
@@ -276,7 +276,11 @@ struct file_system_type *get_fs_type(const char *name)
 	int len = dot ? dot - name : strlen(name);
 
 	fs = __get_fs_type(name, len);
+#ifdef CONFIG_GRKERNSEC_MODHARDEN
+	if (!fs && (___request_module(true, "grsec_modharden_fs", "fs-%.*s", len, name) == 0))
+#else
 	if (!fs && (request_module("fs-%.*s", len, name) == 0)) {
+#endif
 		fs = __get_fs_type(name, len);
 		WARN_ONCE(!fs, "request_module fs-%.*s succeeded, but still no fs?\n", len, name);
 	}
