diff --git a/include/drm/drm_drv.h b/include/drm/drm_drv.h
index 5699f42..dd1ad66 100644
--- a/include/drm/drm_drv.h
+++ b/include/drm/drm_drv.h
@@ -533,9 +533,10 @@ struct drm_driver {
 	int (*dma_quiescent) (struct drm_device *);
 	int (*context_dtor) (struct drm_device *dev, int context);
 	int dev_priv_size;
-};
+} __do_const;
+typedef struct drm_driver __no_const drm_driver_no_const;
 
-__printf(6, 7)
+extern __printf(6, 7) __nocapture(4, 5)
 void drm_dev_printk(const struct device *dev, const char *level,
 		    unsigned int category, const char *function_name,
 		    const char *prefix, const char *format, ...);
