diff --git a/include/linux/cpufreq.h b/include/linux/cpufreq.h
index a5ce0bb..47cbf86 100644
--- a/include/linux/cpufreq.h
+++ b/include/linux/cpufreq.h
@@ -228,6 +228,7 @@ struct freq_attr {
 	ssize_t (*show)(struct cpufreq_policy *, char *);
 	ssize_t (*store)(struct cpufreq_policy *, const char *, size_t count);
 };
+typedef struct freq_attr __no_const freq_attr_no_const;
 
 #define cpufreq_freq_attr_ro(_name)		\
 static struct freq_attr _name =			\
@@ -248,10 +249,11 @@ __ATTR(_name, 0200, NULL, store_##_name)
 struct global_attr {
 	struct attribute attr;
 	ssize_t (*show)(struct kobject *kobj,
-			struct attribute *attr, char *buf);
-	ssize_t (*store)(struct kobject *a, struct attribute *b,
+			struct kobj_attribute *attr, char *buf);
+	ssize_t (*store)(struct kobject *a, struct kobj_attribute *b,
 			 const char *c, size_t count);
 };
+typedef struct global_attr __no_const global_attr_no_const;
 
 #define define_one_global_ro(_name)		\
 static struct global_attr _name =		\
@@ -334,7 +336,7 @@ struct cpufreq_driver {
 	/* platform specific boost support code */
 	bool		boost_enabled;
 	int		(*set_boost)(int state);
-};
+} __do_const;
 
 /* flags */
 #define CPUFREQ_STICKY		(1 << 0)	/* driver isn't removed even if
