diff --git a/include/net/9p/transport.h b/include/net/9p/transport.h
index 5122b5e..598b440 100644
--- a/include/net/9p/transport.h
+++ b/include/net/9p/transport.h
@@ -63,7 +63,7 @@ struct p9_trans_module {
 	int (*zc_request)(struct p9_client *, struct p9_req_t *,
 			  struct iov_iter *, struct iov_iter *, int , int, int);
 	int (*show_options)(struct seq_file *, struct p9_client *);
-};
+} __do_const;
 
 void v9fs_register_trans(struct p9_trans_module *m);
 void v9fs_unregister_trans(struct p9_trans_module *m);
