diff --git a/include/linux/cred.h b/include/linux/cred.h
index f0e70a1..0de3207 100644
--- a/include/linux/cred.h
+++ b/include/linux/cred.h
@@ -203,6 +203,9 @@ static inline void validate_creds_for_do_exit(struct task_struct *tsk)
 static inline void validate_process_creds(void)
 {
 }
+static inline void validate_task_creds(struct task_struct *task)
+{
+}
 #endif
 
 static inline bool cap_ambient_invariant_ok(const struct cred *cred)
@@ -347,6 +350,7 @@ static inline void put_cred(const struct cred *_cred)
 
 #define task_uid(task)		(task_cred_xxx((task), uid))
 #define task_euid(task)		(task_cred_xxx((task), euid))
+#define task_securebits(task)	(task_cred_xxx((task), securebits))
 
 #define current_cred_xxx(xxx)			\
 ({						\
