diff --git a/include/sound/pcm.h b/include/sound/pcm.h
index af1fb37..0432863 100644
--- a/include/sound/pcm.h
+++ b/include/sound/pcm.h
@@ -1072,7 +1072,7 @@ void snd_pcm_set_ops(struct snd_pcm * pcm, int direction,
 void snd_pcm_set_sync(struct snd_pcm_substream *substream);
 int snd_pcm_lib_ioctl(struct snd_pcm_substream *substream,
 		      unsigned int cmd, void *arg);                      
-void snd_pcm_period_elapsed(struct snd_pcm_substream *substream);
+void snd_pcm_period_elapsed(void *_substream);
 snd_pcm_sframes_t __snd_pcm_lib_xfer(struct snd_pcm_substream *substream,
 				     void *buf, bool interleaved,
 				     snd_pcm_uframes_t frames, bool in_kernel);
