diff --git a/include/net/inet_connection_sock.h b/include/net/inet_connection_sock.h
index 197a30d..677cd909 100644
--- a/include/net/inet_connection_sock.h
+++ b/include/net/inet_connection_sock.h
@@ -63,7 +63,7 @@ struct inet_connection_sock_af_ops {
 #endif
 	void	    (*addr2sockaddr)(struct sock *sk, struct sockaddr *);
 	void	    (*mtu_reduced)(struct sock *sk);
-};
+} __do_const;
 
 /** inet_connection_sock - INET connection oriented sock
  *
