diff --git a/include/linux/netfilter/ipset/ip_set.h b/include/linux/netfilter/ipset/ip_set.h
index 83b9a2e..5266f3b 100644
--- a/include/linux/netfilter/ipset/ip_set.h
+++ b/include/linux/netfilter/ipset/ip_set.h
@@ -95,8 +95,8 @@ struct ip_set_ext_type {
 extern const struct ip_set_ext_type ip_set_extensions[];
 
 struct ip_set_counter {
-	atomic64_t bytes;
-	atomic64_t packets;
+	atomic64_unchecked_t bytes;
+	atomic64_unchecked_t packets;
 };
 
 struct ip_set_comment_rcu {
