diff --git a/init/do_mounts.h b/init/do_mounts.h
index 067af1d..b535547 100644
--- a/init/do_mounts.h
+++ b/init/do_mounts.h
@@ -15,14 +15,14 @@ extern int root_mountflags;
 
 static inline int create_dev(char *name, dev_t dev)
 {
-	sys_unlink(name);
-	return sys_mknod(name, S_IFBLK|0600, new_encode_dev(dev));
+	sys_unlink((char __force_user *)name);
+	return sys_mknod((char __force_user *)name, S_IFBLK|0600, new_encode_dev(dev));
 }
 
 static inline u32 bstat(char *name)
 {
 	struct kstat stat;
-	if (vfs_stat(name, &stat) != 0)
+	if (vfs_stat((char __force_user *)name, (struct kstat __force_user *)&stat) != 0)
 		return 0;
 	if (!S_ISBLK(stat.mode))
 		return 0;
