diff --git a/include/linux/average.h b/include/linux/average.h
index 7ddaf34..aa614e3 100644
--- a/include/linux/average.h
+++ b/include/linux/average.h
@@ -57,7 +57,7 @@
 		BUILD_BUG_ON((_precision) > 30);			\
 		BUILD_BUG_ON_NOT_POWER_OF_2(_weight_rcp);		\
 									\
-		ACCESS_ONCE(e->internal) = internal ?			\
+		ACCESS_ONCE_RW(e->internal) = internal ?		\
 			(((internal << weight_rcp) - internal) +	\
 				(val << precision)) >> weight_rcp :	\
 			(val << precision);				\
