diff --git a/kernel/audit.c b/kernel/audit.c
index f1ca116..7f396f8 100644
--- a/kernel/audit.c
+++ b/kernel/audit.c
@@ -146,7 +146,7 @@ u32		audit_sig_sid = 0;
    3) suppressed due to audit_rate_limit
    4) suppressed due to audit_backlog_limit
 */
-static atomic_t	audit_lost = ATOMIC_INIT(0);
+static atomic_unchecked_t	audit_lost = ATOMIC_INIT(0);
 
 /* Hash for inode-based rules */
 struct list_head audit_inode_hash[AUDIT_INODE_BUCKETS];
@@ -256,7 +256,7 @@ void audit_log_lost(const char *message)
 	unsigned long		now;
 	int			print;
 
-	atomic_inc(&audit_lost);
+	atomic_inc_unchecked(&audit_lost);
 
 	print = (audit_failure == AUDIT_FAIL_PANIC || !audit_rate_limit);
 
@@ -273,7 +273,7 @@ void audit_log_lost(const char *message)
 	if (print) {
 		if (printk_ratelimit())
 			pr_warn("audit_lost=%u audit_rate_limit=%u audit_backlog_limit=%u\n",
-				atomic_read(&audit_lost),
+				atomic_read_unchecked(&audit_lost),
 				audit_rate_limit,
 				audit_backlog_limit);
 		audit_panic(message);
@@ -1166,7 +1166,7 @@ static int audit_receive_msg(struct sk_buff *skb, struct nlmsghdr *nlh)
 		s.pid			= auditd_pid_vnr();
 		s.rate_limit		= audit_rate_limit;
 		s.backlog_limit		= audit_backlog_limit;
-		s.lost			= atomic_read(&audit_lost);
+		s.lost			= atomic_read_unchecked(&audit_lost);
 		s.backlog		= skb_queue_len(&audit_queue);
 		s.feature_bitmap	= AUDIT_FEATURE_BITMAP_ALL;
 		s.backlog_wait_time	= audit_backlog_wait_time;
@@ -1261,7 +1261,7 @@ static int audit_receive_msg(struct sk_buff *skb, struct nlmsghdr *nlh)
 				return err;
 		}
 		if (s.mask == AUDIT_STATUS_LOST) {
-			u32 lost = atomic_xchg(&audit_lost, 0);
+			u32 lost = atomic_xchg_unchecked(&audit_lost, 0);
 
 			audit_log_config_change("lost", 0, lost, 1);
 			return lost;
@@ -1455,7 +1455,7 @@ static void __net_exit audit_net_exit(struct net *net)
 	netlink_kernel_release(aunet->sk);
 }
 
-static struct pernet_operations audit_net_ops __net_initdata = {
+static struct pernet_operations audit_net_ops __net_initconst = {
 	.init = audit_net_init,
 	.exit = audit_net_exit,
 	.id = &audit_net_id,
