diff --git a/net/mpls/af_mpls.c b/net/mpls/af_mpls.c
index 1309e2c..6e543c6 100644
--- a/net/mpls/af_mpls.c
+++ b/net/mpls/af_mpls.c
@@ -1180,7 +1180,7 @@ static int mpls_dev_sysctl_register(struct net_device *dev,
 {
 	char path[sizeof("net/mpls/conf/") + IFNAMSIZ];
 	struct net *net = dev_net(dev);
-	struct ctl_table *table;
+	ctl_table_no_const *table;
 	int i;
 
 	table = kmemdup(&mpls_dev_table, sizeof(mpls_dev_table), GFP_KERNEL);
@@ -1622,7 +1622,7 @@ static int mpls_platform_labels(struct ctl_table *table, int write,
 	struct net *net = table->data;
 	int platform_labels = net->mpls.platform_labels;
 	int ret;
-	struct ctl_table tmp = {
+	ctl_table_no_const tmp = {
 		.procname	= table->procname,
 		.data		= &platform_labels,
 		.maxlen		= sizeof(int),
@@ -2149,7 +2149,7 @@ static const struct ctl_table mpls_table[] = {
 
 static int mpls_net_init(struct net *net)
 {
-	struct ctl_table *table;
+	ctl_table_no_const *table;
 	int i;
 
 	net->mpls.platform_labels = 0;
