diff --git a/net/8021q/vlan.c b/net/8021q/vlan.c
index f2531ad..5792672 100644
--- a/net/8021q/vlan.c
+++ b/net/8021q/vlan.c
@@ -496,7 +496,7 @@ static int vlan_device_event(struct notifier_block *unused, unsigned long event,
 	return NOTIFY_DONE;
 }
 
-static struct notifier_block vlan_notifier_block __read_mostly = {
+static struct notifier_block vlan_notifier_block = {
 	.notifier_call = vlan_device_event,
 };
 
