diff --git a/net/sunrpc/cache.c b/net/sunrpc/cache.c
index 79d55d9..484f913 100644
--- a/net/sunrpc/cache.c
+++ b/net/sunrpc/cache.c
@@ -1617,7 +1617,7 @@ static int create_cache_proc_entries(struct cache_detail *cd, struct net *net)
 	struct sunrpc_net *sn;
 
 	sn = net_generic(net, sunrpc_net_id);
-	cd->procfs = proc_mkdir(cd->name, sn->proc_net_rpc);
+	cd->procfs = proc_mkdir_restrict(cd->name, sn->proc_net_rpc);
 	if (cd->procfs == NULL)
 		goto out_nomem;
 
