diff --git a/net/dsa/slave.c b/net/dsa/slave.c
index 3ff9d97..a17eb98 100644
--- a/net/dsa/slave.c
+++ b/net/dsa/slave.c
@@ -861,7 +861,7 @@ static int dsa_slave_setup_tc(struct net_device *dev, u32 handle,
 	}
 }
 
-void dsa_cpu_port_ethtool_init(struct ethtool_ops *ops)
+void dsa_cpu_port_ethtool_init(ethtool_ops_no_const *ops)
 {
 	ops->get_sset_count = dsa_cpu_port_get_sset_count;
 	ops->get_ethtool_stats = dsa_cpu_port_get_ethtool_stats;
@@ -1489,7 +1489,7 @@ static int dsa_slave_switchdev_event(struct notifier_block *unused,
 	return NOTIFY_BAD;
 }
 
-static struct notifier_block dsa_slave_nb __read_mostly = {
+static struct notifier_block dsa_slave_nb = {
 	.notifier_call  = dsa_slave_netdevice_event,
 };
 
