diff --git a/net/sunrpc/rpcb_clnt.c b/net/sunrpc/rpcb_clnt.c
index 5b30603..d49e23a 100644
--- a/net/sunrpc/rpcb_clnt.c
+++ b/net/sunrpc/rpcb_clnt.c
@@ -843,9 +843,10 @@ static void rpcb_getport_done(struct rpc_task *child, void *data)
  * XDR functions for rpcbind
  */
 
-static void rpcb_enc_mapping(struct rpc_rqst *req, struct xdr_stream *xdr,
+static void rpcb_enc_mapping(void *_req, struct xdr_stream *xdr,
 			     const void *data)
 {
+	struct rpc_rqst *req = _req;
 	const struct rpcbind_args *rpcb = data;
 	__be32 *p;
 
@@ -864,6 +865,7 @@ static void rpcb_enc_mapping(struct rpc_rqst *req, struct xdr_stream *xdr,
 static int rpcb_dec_getport(struct rpc_rqst *req, struct xdr_stream *xdr,
 			    void *data)
 {
+	struct rpc_rqst *req = _req;
 	struct rpcbind_args *rpcb = data;
 	unsigned long port;
 	__be32 *p;
@@ -887,6 +889,7 @@ static int rpcb_dec_getport(struct rpc_rqst *req, struct xdr_stream *xdr,
 static int rpcb_dec_set(struct rpc_rqst *req, struct xdr_stream *xdr,
 			void *data)
 {
+	struct rpc_rqst *req = _req;
 	unsigned int *boolp = data;
 	__be32 *p;
 
@@ -920,9 +923,10 @@ static void encode_rpcb_string(struct xdr_stream *xdr, const char *string,
 	xdr_encode_opaque(p, string, len);
 }
 
-static void rpcb_enc_getaddr(struct rpc_rqst *req, struct xdr_stream *xdr,
+static void rpcb_enc_getaddr(void *_req, struct xdr_stream *xdr,
 			     const void *data)
 {
+	struct rpc_rqst *req = _req;
 	const struct rpcbind_args *rpcb = data;
 	__be32 *p;
 
@@ -941,9 +945,10 @@ static void rpcb_enc_getaddr(struct rpc_rqst *req, struct xdr_stream *xdr,
 	encode_rpcb_string(xdr, rpcb->r_owner, RPCB_MAXOWNERLEN);
 }
 
-static int rpcb_dec_getaddr(struct rpc_rqst *req, struct xdr_stream *xdr,
+static int rpcb_dec_getaddr(void *_req, struct xdr_stream *xdr,
 			    void *data)
 {
+	struct rpc_rqst *req = _req;
 	struct rpcbind_args *rpcb = data;
 	struct sockaddr_storage address;
 	struct sockaddr *sap = (struct sockaddr *)&address;
