diff --git a/scripts/gcc-plugins/sancov_plugin.c b/scripts/gcc-plugins/sancov_plugin.c
index 0f98634..59967ab 100644
--- a/scripts/gcc-plugins/sancov_plugin.c
+++ b/scripts/gcc-plugins/sancov_plugin.c
@@ -62,6 +62,7 @@ static unsigned int sancov_execute(void)
 
 #include "gcc-generate-gimple-pass.h"
 
+#if BUILDING_GCC_VERSION < 6000
 static void sancov_start_unit(void __unused *gcc_data, void __unused *user_data)
 {
 	tree leaf_attr, nothrow_attr;
@@ -85,6 +86,7 @@ static void sancov_start_unit(void __unused *gcc_data, void __unused *user_data)
 	decl_attributes(&sancov_fndecl, leaf_attr, 0);
 #endif
 }
+#endif
 
 __visible int plugin_init(struct plugin_name_args *plugin_info, struct plugin_gcc_version *version)
 {
@@ -113,7 +115,7 @@ __visible int plugin_init(struct plugin_name_args *plugin_info, struct plugin_gc
 #endif
 
 	if (!plugin_default_version_check(version, &gcc_version)) {
-		error(G_("incompatible gcc/plugin versions"));
+		error_gcc_version(version);
 		return 1;
 	}
 
