diff --git a/security/apparmor/policy_ns.c b/security/apparmor/policy_ns.c
index 93d1826..07a4a79 100644
--- a/security/apparmor/policy_ns.c
+++ b/security/apparmor/policy_ns.c
@@ -116,7 +116,7 @@ static struct aa_ns *alloc_ns(const char *prefix, const char *name)
 	/* ns and ns->unconfined share ns->unconfined refcount */
 	ns->unconfined->ns = ns;
 
-	atomic_set(&ns->uniq_null, 0);
+	atomic_set_unchecked(&ns->uniq_null, 0);
 
 	aa_labelset_init(&ns->labels);
 
