diff --git a/security/apparmor/policy.c b/security/apparmor/policy.c
index 179e68d..48bd588 100644
--- a/security/apparmor/policy.c
+++ b/security/apparmor/policy.c
@@ -354,7 +354,7 @@ struct aa_profile *aa_new_null_profile(struct aa_profile *parent, bool hat,
 	if (!name)
 		return NULL;
 	sprintf(name, "%s//null-%x", parent->base.hname,
-		atomic_inc_return(&parent->ns->uniq_null));
+		atomic_inc_return_unchecked(&parent->ns->uniq_null));
 
 name:
 	/* lookup to see if this is a dup creation */
